File: java/PairSumRemoval.java
import java.io.*;
import java.util.HashMap;
import java.util.Scanner;

public class PairSumRemoval {

    public static int maxOperations(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int count = 0;

        for (int num : nums) {
            int complement = k - num;
            if (map.getOrDefault(complement, 0) > 0) {
                count++;
                map.put(complement, map.get(complement) - 1);
            } else {
                map.put(num, map.getOrDefault(num, 0) + 1);
            }
        }

        return count;
    }

    public static void main(String[] args) {
        int[] arr = { 3, 1, 3, 4, 3 };
        int k = 6;
        System.out.println(maxOperations(arr, k));
    }
}

File: medium/arrays/pairSumRemoval.go
package main

import "fmt"

func main() {
	arr := []int{1, 2, 3, 4}
	k := 5

	fmt.Println(maxOperations(arr, k))
}

func maxOperations(arr []int, k int) int {
	done := map[int]int{}
	count := 0
	for _, num := range arr {
		fmt.Println(done)
		complement := k - num
		fmt.Println(complement)
		if found := done[complement]; found > 0 {
			count++
			done[complement]--
		} else {
			done[num]++
		}
	}
	return count
}
